report/topology.go

```diff
@@ -195,6 +197,100 @@ func (c Counters) Copy() Counters {
 	return result
 }
 
+// Sets is a string->set-of-strings map.
+type Sets map[string]StringSet
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (s Sets) Merge(other Sets) Sets {
+	result := s.Copy()
+	for k, v := range other {
+		result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (s Sets) Copy() Sets {
+	result := Sets{}
+	for k, v := range s {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// StringSet is a sorted set of unique strings. Clients must use the Add
+// method to add strings.
+type StringSet []string
```

Comment:
Though you could implement the methods on a map so that the zero type is still valid:
```Go
func (s set) add(v string) set {
	if s == nil {
		s = make(map[string]struct{})
	}
	s[v] = struct{}{}
	return s
}
```